File: Catalog REST/src/main/java/cz/vhromada/catalog/rest/controller/EpisodeController.java
package cz.vhromada.catalog.rest.controller;

import cz.vhromada.catalog.facade.EpisodeFacade;
import cz.vhromada.catalog.facade.to.EpisodeTO;
import cz.vhromada.catalog.facade.to.SeasonTO;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;

/**
 * A class represents controller for episodes.
 *
 * @author Vladimir Hromada
 */
@Controller("episodeController")
@RequestMapping("/series/{serieId}/seasons/{seasonId}/episodes")
public class EpisodeController extends JsonController {

    @Autowired
    @Qualifier("episodeFacade")
    private EpisodeFacade episodeFacade;

    /**
     * Returns episode with ID or null if there isn't such episode.
     *
     * @param serieId   serie ID
     * @param seasonId  season ID
     * @param episodeId episode ID
     * @return episode with ID or null if there isn't such episode
     * @throws IllegalArgumentException if ID is null
     */
    @RequestMapping(value = "/{episodeId}", method = RequestMethod.GET)
    @ResponseBody
    public String getEpisode(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") @SuppressWarnings("unused") final Integer seasonId, @PathVariable("episodeId") final Integer episodeId) {
        return serialize(episodeFacade.getEpisode(episodeId));
    }

    /**
     * Adds episode. Sets new ID and position.
     *
     * @param serieId  serie ID
     * @param seasonId season ID
     * @param episode  episode
     * @throws IllegalArgumentException                                  if episode is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID isn't null
     *                                                                   or name is null
     *                                                                   or name is empty string
     *                                                                   or length of episode is negative value
     *                                                                   or note is null
     *                                                                   or season is null
     *                                                                   or season ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if season doesn't exist in data storage
     */
    @RequestMapping(value = "/add", method = RequestMethod.POST)
    @ResponseBody
    public void add(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") @SuppressWarnings("unused") final Integer seasonId, final String episode) {
        episodeFacade.add(deserialize(episode, EpisodeTO.class));
    }

    /**
     * Updates episode.
     *
     * @param serieId  serie ID
     * @param seasonId season ID
     * @param episode  new value of episode
     * @throws IllegalArgumentException                                  if episode is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     *                                                                   or name is null
     *                                                                   or name is empty string
     *                                                                   or length of episode is negative value
     *                                                                   or note is null
     *                                                                   or season is null
     *                                                                   or season ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if episode doesn't exist in data storage
     *                                                                   or season doesn't exist in data storage
     */
    @RequestMapping(value = "/update", method = RequestMethod.POST)
    @ResponseBody
    public void update(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") @SuppressWarnings("unused") final Integer seasonId, final String episode) {
        episodeFacade.update(deserialize(episode, EpisodeTO.class));
    }

    /**
     * Removes episode.
     *
     * @param serieId  serie ID
     * @param seasonId season ID
     * @param episode  episode
     * @throws IllegalArgumentException                                  if episode is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if episode doesn't exist in data storage
     */
    @RequestMapping(value = "/remove", method = RequestMethod.POST)
    @ResponseBody
    public void remove(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") @SuppressWarnings("unused") final Integer seasonId, final String episode) {
        episodeFacade.remove(deserialize(episode, EpisodeTO.class));
    }

    /**
     * Duplicates episode.
     *
     * @param serieId  serie ID
     * @param seasonId season ID
     * @param episode  episode
     * @throws IllegalArgumentException                                  if episode is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if episode doesn't exist in data storage
     */
    @RequestMapping(value = "/duplicate", method = RequestMethod.POST)
    @ResponseBody
    public void duplicate(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") @SuppressWarnings("unused") final Integer seasonId, final String episode) {
        episodeFacade.duplicate(deserialize(episode, EpisodeTO.class));
    }

    /**
     * Moves episode in list one position up.
     *
     * @param serieId  serie ID
     * @param seasonId season ID
     * @param episode  episode
     * @throws IllegalArgumentException                                  if episode is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     *                                                                   or episode can't be moved up
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if episode doesn't exist in data storage
     */
    @RequestMapping(value = "/moveUp", method = RequestMethod.POST)
    @ResponseBody
    public void moveUp(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") @SuppressWarnings("unused") final Integer seasonId, final String episode) {
        episodeFacade.moveUp(deserialize(episode, EpisodeTO.class));
    }

    /**
     * Moves episode in list one position down.
     *
     * @param serieId  serie ID
     * @param seasonId season ID
     * @param episode  episode
     * @throws IllegalArgumentException                                  if episode is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     *                                                                   or episode can't be moved down
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if episode doesn't exist in data storage
     */
    @RequestMapping(value = "/moveDown", method = RequestMethod.POST)
    @ResponseBody
    public void moveDown(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") @SuppressWarnings("unused") final Integer seasonId, final String episode) {
        episodeFacade.moveDown(deserialize(episode, EpisodeTO.class));
    }

    /**
     * Returns true if episode exists.
     *
     * @param serieId  serie ID
     * @param seasonId season ID
     * @param episode  episode
     * @return true if episode exists
     * @throws IllegalArgumentException                              if episode is null
     * @throws cz.vhromada.validators.exceptions.ValidationException if ID is null
     */
    @RequestMapping(value = "/exists", method = RequestMethod.GET)
    @ResponseBody
    public String exists(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") @SuppressWarnings("unused") final Integer seasonId, final String episode) {
        return serialize(episodeFacade.exists(deserialize(episode, EpisodeTO.class)));
    }

    /**
     * Returns list of episodes for specified season.
     *
     * @param serieId  serie ID
     * @param seasonId season ID
     * @return list of episodes for specified season
     * @throws IllegalArgumentException                                  if season is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if season doesn't exist in data storage
     */
    @RequestMapping(value = { "", "/", "list" }, method = RequestMethod.GET)
    @ResponseBody
    public String findEpisodesBySeason(@PathVariable("serieId") @SuppressWarnings("unused") final Integer serieId,
            @PathVariable("seasonId") final Integer seasonId) {
        final SeasonTO season = new SeasonTO();
        season.setId(seasonId);

        return serialize(episodeFacade.findEpisodesBySeason(season));
    }

}

File: catalog-rest/src/main/kotlin/com/github/vhromada/catalog/rest/controller/EpisodeController.kt
package com.github.vhromada.catalog.rest.controller

import com.github.vhromada.catalog.entity.Episode
import com.github.vhromada.catalog.facade.EpisodeFacade
import com.github.vhromada.catalog.facade.SeasonFacade
import com.github.vhromada.catalog.facade.ShowFacade
import com.github.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for episodes.
 *
 * @author Vladimir Hromada
 */
@RestController("episodeController")
@RequestMapping("/catalog/shows/{showId}/seasons/{seasonId}/episodes")
class EpisodeController(
    private val episodeFacade: EpisodeFacade,
    private val showFacade: ShowFacade,
    private val seasonFacade: SeasonFacade
) : AbstractController() {

    /**
     * Returns episode with ID.
     * <br></br>
     * Validation errors:
     *
     *  * Show doesn't exist in data storage
     *  * Season doesn't exist in data storage
     *  * Episode doesn't exist in data storage
     *
     * @param showId    show ID
     * @param seasonId  season ID
     * @param episodeId episode ID
     * @return episode with ID
     */
    @GetMapping("/{episodeId}")
    fun getEpisode(
        @PathVariable("showId") showId: Int,
        @PathVariable("seasonId") seasonId: Int,
        @PathVariable("episodeId") episodeId: Int
    ): Episode {
        processResult(result = showFacade.get(id = showId))
        processResult(result = seasonFacade.get(id = seasonId))
        return processResult(result = episodeFacade.get(id = episodeId))!!
    }

    /**
     * Adds episode. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * Show doesn't exist in data storage
     *  * Season doesn't exist in data storage
     *  * Episode ID isn't null
     *  * Episode position isn't null
     *  * Number of episode isn't positive number
     *  * Name is null
     *  * Name is empty string
     *  * Length of episode is negative value
     *  * Note is null
     *
     * @param showId   show ID
     * @param seasonId season ID
     * @param episode  episode
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(
        @PathVariable("showId") showId: Int,
        @PathVariable("seasonId") seasonId: Int,
        @RequestBody episode: Episode
    ) {
        processResult(result = showFacade.get(id = showId))
        processResult(result = episodeFacade.add(parent = seasonId, data = episode))
    }

    /**
     * Updates episode.
     * <br></br>
     * Validation errors:
     *
     *  * Show doesn't exist in data storage
     *  * Season doesn't exist in data storage
     *  * ID is null
     *  * Position is null
     *  * Number of episode isn't positive number
     *  * Name is null
     *  * Name is empty string
     *  * Length of episode is negative value
     *  * Note is null
     *  * Episode doesn't exist in data storage
     *
     * @param showId   show ID
     * @param seasonId season ID
     * @param episode  new value of episode
     */
    @PostMapping("/update")
    fun update(
        @PathVariable("showId") showId: Int,
        @PathVariable("seasonId") seasonId: Int,
        @RequestBody episode: Episode
    ) {
        processResult(result = showFacade.get(id = showId))
        processResult(result = seasonFacade.get(id = seasonId))
        processResult(result = episodeFacade.update(data = episode))
    }

    /**
     * Removes episode.
     * <br></br>
     * Validation errors:
     *
     *  * Show doesn't exist in data storage
     *  * Season doesn't exist in data storage
     *  * Episode doesn't exist in data storage
     *
     * @param showId   show ID
     * @param seasonId season ID
     * @param id       ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(
        @PathVariable("showId") showId: Int,
        @PathVariable("seasonId") seasonId: Int,
        @PathVariable("id") id: Int
    ) {
        processResult(result = showFacade.get(id = showId))
        processResult(result = seasonFacade.get(id = seasonId))
        processResult(result = episodeFacade.remove(id = id))
    }

    /**
     * Duplicates episode.
     * <br></br>
     * Validation errors:
     *
     *  * Show doesn't exist in data storage
     *  * Season doesn't exist in data storage
     *  * Episode doesn't exist in data storage
     *
     * @param showId   show ID
     * @param seasonId season ID
     * @param id       ID
     */
    @PostMapping("/duplicate/{id}")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(
        @PathVariable("showId") showId: Int,
        @PathVariable("seasonId") seasonId: Int,
        @PathVariable("id") id: Int
    ) {
        processResult(result = showFacade.get(id = showId))
        processResult(result = seasonFacade.get(id = seasonId))
        processResult(result = episodeFacade.duplicate(id = id))
    }

    /**
     * Moves episode in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * Show doesn't exist in data storage
     *  * Season doesn't exist in data storage
     *  * Episode can't be moved up
     *  * Episode doesn't exist in data storage
     *
     * @param showId   show ID
     * @param seasonId season ID
     * @param id       ID
     */
    @PostMapping("/moveUp/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(
        @PathVariable("showId") showId: Int,
        @PathVariable("seasonId") seasonId: Int,
        @PathVariable("id") id: Int
    ) {
        processResult(result = showFacade.get(id = showId))
        processResult(result = seasonFacade.get(id = seasonId))
        processResult(result = episodeFacade.moveUp(id = id))
    }

    /**
     * Moves episode in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * Show doesn't exist in data storage
     *  * Season doesn't exist in data storage
     *  * Episode can't be moved up
     *  * Episode doesn't exist in data storage
     *
     * @param showId   show ID
     * @param seasonId season ID
     * @param id       ID
     */
    @PostMapping("/moveDown/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(
        @PathVariable("showId") showId: Int,
        @PathVariable("seasonId") seasonId: Int,
        @PathVariable("id") id: Int
    ) {
        processResult(result = showFacade.get(id = showId))
        processResult(result = seasonFacade.get(id = seasonId))
        processResult(result = episodeFacade.moveDown(id = id))
    }

    /**
     * Returns list of episodes for specified season.
     * <br></br>
     * Validation errors:
     *
     *  * Show doesn't exist in data storage
     *  * Season doesn't exist in data storage
     *
     * @param showId   show ID
     * @param seasonId season ID
     * @return list of episodes for specified season
     */
    @GetMapping
    fun findEpisodesBySeason(
        @PathVariable("showId") showId: Int,
        @PathVariable("seasonId") seasonId: Int
    ): List<Episode> {
        processResult(result = showFacade.get(id = showId))
        return processResult(result = episodeFacade.find(parent = seasonId))!!
    }

}
